File: 68. Text Justification/68. Text Justification.dart
class Solution {
  List<String> fullJustify(List<String> words, int maxWidth) {
    List<String> result = [];
    int start = 0, len = 0;
    
    for (int i = 0; i <= words.length; i++) {
      if (i == words.length || len + words[i].length + i - start > maxWidth) {
        int spaces = maxWidth - len;
        int gaps = i - start - 1;
        
        StringBuffer sb = StringBuffer();
        for (int j = start; j < i; j++) {
          sb.write(words[j]);
          if (j == i - 1) break;
          
          int extraSpaces = (i == words.length || gaps == 0) ? 1 : (spaces ~/ gaps) + (spaces % gaps > 0 ? 1 : 0);
          sb.write(' ' * extraSpaces);
          spaces -= extraSpaces;
          gaps--;
        }
        
        sb.write(' ' * spaces);
        result.add(sb.toString());
        
        start = i;
        len = 0;
      }
      if (i < words.length) len += words[i].length;
    }
    
    return result;
  }
}

File: 68. Text Justification/68. Text Justification.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<String> fullJustify(String[] words, int maxWidth) {
        List<String> result = new ArrayList<>();
        int n = words.length;
        int i = 0;
        
        while (i < n) {
            int j = i, lineLength = 0;
            
            // Find words that fit in this line
            while (j < n && lineLength + words[j].length() + (j - i) <= maxWidth) {
                lineLength += words[j].length();
                j++;
            }
            
            int spaces = maxWidth - lineLength;
            int gaps = j - i - 1;
            
            StringBuilder line = new StringBuilder();
            
            // Last line or line with only one word
            if (j == n || gaps == 0) {
                for (int k = i; k < j; k++) {
                    line.append(words[k]);
                    if (k < j - 1) {
                        line.append(" ");
                        spaces--;
                    }
                }
                line.append(" ".repeat(spaces));
            } 
            else {
                int spacePerGap = spaces / gaps;
                int extraSpaces = spaces % gaps;
                
                for (int k = i; k < j - 1; k++) {
                    line.append(words[k]);
                    line.append(" ".repeat(spacePerGap));
                    if (extraSpaces > 0) {
                        line.append(" ");
                        extraSpaces--;
                    }
                }
                line.append(words[j-1]);
            }
            
            result.add(line.toString());
            i = j;
        }
        
        return result;
    }
}